Unable to build with MSVC 19.32 with `/std:latest` and `/Zc:__cplusplus` : `fatal error C1001: Internal compiler error`

### What version of gRPC and what language are you using?

v1.48.0

### What operating system (Linux, Windows,...) and version?

Windows 11

### What runtime / compiler are you using (e.g. python version or version of gcc)

MSVC:  19.32.31332.0

### What did you do?

Building with cmake.

```bash
'C:/Program Files/CMake/bin/cmake.exe' '../../../../a/cmake-toolset/cmake-toolset/test/third_party/packages/grpc-v1.48.0' '-Wno-dev' '-DCMAKE_INSTALL_PREFIX=D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc' '-G' 'Visual Studio 17 2022' '-DCMAKE_EXE_LINKER_FLAGS=/machine:x64' '-DCMAKE_MODULE_LINKER_FLAGS=/machine:x64' '-DCMAKE_SHARED_LINKER_FLAGS=/machine:x64' '-DCMAKE_STATIC_LINKER_FLAGS=/machine:x64' '-DCMAKE_SYSTEM_VERSION=10.0.22621.0' '-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON' '-DCMAKE_EXPORT_PACKAGE_REGISTRY=OFF' '-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded$<$<CONFIG:Debug>:Debug>$<$<NOT:$<STREQUAL:,static>>:DLL>' '-DCMAKE_C_FLAGS= /DWIN32 /D_WINDOWS /nologo /utf-8 /MP' '-DCMAKE_C_FLAGS_DEBUG= /Zi /Ob0 /Od /RTC1 /D_DEBUG /Z7' '-DCMAKE_C_FLAGS_RELEASE= /O2 /Ob2 /DNDEBUG /Oi /Z7 /Gy' '-DCMAKE_C_FLAGS_RELWITHDEBINFO= /Zi /O2 /Ob1 /DNDEBUG /Oi /Z7 /Gy' '-DCMAKE_C_FLAGS_MINSIZEREL= /O1 /Ob1 /DNDEBUG /Ox /Z7 /Gy' '-DCMAKE_C_STANDARD_LIBRARIES=kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib bcrypt.lib dinput8.lib imm32.lib iphlpapi.lib psapi.lib secur32.lib setupapi.lib shlwapi.lib strmiids.lib version.lib vfw32.lib winmm.lib wsock32.lib userenv.lib Ws2_32.lib wldap32.lib crypt32.lib' '-DCMAKE_CXX_FLAGS= /DWIN32 /D_WINDOWS /EHsc /nologo /utf-8 /MP /Zc:__cplusplus /GR' '-DCMAKE_CXX_FLAGS_DEBUG= /Zi /Ob0 /Od /RTC1 /D_DEBUG /Z7' '-DCMAKE_CXX_FLAGS_RELEASE= /O2 /Ob2 /DNDEBUG /Oi /Z7 /Gy' '-DCMAKE_CXX_FLAGS_RELWITHDEBINFO= /Zi /O2 /Ob1 /DNDEBUG /Oi /Z7 /Gy' '-DCMAKE_CXX_FLAGS_MINSIZEREL= /O1 /Ob1 /DNDEBUG /Ox /Z7 /Gy' '-DCMAKE_CXX_STANDARD_LIBRARIES=kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib bcrypt.lib dinput8.lib imm32.lib iphlpapi.lib psapi.lib secur32.lib setupapi.lib shlwapi.lib strmiids.lib version.lib vfw32.lib winmm.lib wsock32.lib userenv.lib Ws2_32.lib wldap32.lib crypt32.lib' '-A' 'x64' '-DCMAKE_POLICY_DEFAULT_CMP0091=NEW' '-DCMAKE_BUILD_TYPE=Release' '-DCMAKE_C_STANDARD=17' '-DCMAKE_CXX_STANDARD=23' '-DCMAKE_POLICY_DEFAULT_CMP0075=NEW' '-DCMAKE_POSITION_INDEPENDENT_CODE=ON' '-DCMAKE_FIND_ROOT_PATH=D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/lib64/cmake;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/lib/cmake;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/cmake;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/share;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/share/cmake' '-DCMAKE_PREFIX_PATH=D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/lib64/cmake;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/lib/cmake;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/cmake;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/share;D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/share/cmake' '-DCMAKE_POSITION_INDEPENDENT_CODE=ON' '-DgRPC_INSTALL=ON' '-DBUILD_SHARED_LIBS=OFF' '-DCMAKE_BUILD_TYPE=Release' '-DgRPC_INSTALL_CSHARP_EXT=OFF' '-DgRPC_GFLAGS_PROVIDER=none' '-DgRPC_BENCHMARK_PROVIDER=none' '-DgRPC_BUILD_TESTS=OFF' '-DgRPC_BUILD_CODEGEN=ON' '-DgRPC_MSVC_STATIC_RUNTIME=OFF' '-DgRPC_ABSL_PROVIDER=package' '-DgRPC_CARES_PROVIDER=package' '-DgRPC_PROTOBUF_PROVIDER=package' '-DgRPC_PROTOBUF_PACKAGE_TYPE=CONFIG' '-DgRPC_RE2_PROVIDER=package' '-DgRPC_ZLIB_PROVIDER=package' '-DgRPC_SSL_PROVIDER=package' '-D_gRPC_PROTOBUF_PROTOC_EXECUTABLE=D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/bin/protoc.exe' '-DProtobuf_PROTOC_EXECUTABLE=D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/bin/protoc.exe' '-DPROTOBUF_PROTOC_EXECUTABLE=D:/a/cmake-toolset/cmake-toolset/test/third_party/install/windows-amd64-msvc/bin/protoc.exe'
```

### What did you expect to see?

Build success.

### What did you see instead?

```
D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src/core/lib/promise/detail/basic_seq.h(481,1): fatal error C1001: Internal compiler error. [D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
  (compiler file 'msc1.cpp', line 1693)
   To work around this problem, try simplifying or changing the program near the locations listed above.
  If possible please provide a repro here: https://developercommunity.visualstudio.com/
  Please choose the Technical Support command on the Visual C++
   Help menu, or open the Technical Support help file for more information (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\security\credentials\composite\composite_credentials.cc)
D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src/core/lib/promise/detail/basic_seq.h(466): message : while compiling class template member function 'absl::lts_20211102::variant<grpc_core::Pending,T> grpc_core::promise_detail::BasicSeqIter<grpc_core::promise_detail::TrySeqTraits,grpc_composite_call_credentials::GetRequestMetadata::<lambda_1>,grpc_core::MetadataHandle<grpc_core::ServerMetadata>,std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>::PollNonEmpty(void)' [D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
          with
          [
              T=absl::lts_20211102::StatusOr<grpc_core::ClientMetadataHandle>,
              _Ty=grpc_core::RefCountedPtr<grpc_call_credentials>
          ] (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\security\credentials\composite\composite_credentials.cc)
D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src/core/lib/promise/detail/basic_seq.h(462): message : see reference to function template instantiation 'absl::lts_20211102::variant<grpc_core::Pending,T> grpc_core::promise_detail::BasicSeqIter<grpc_core::promise_detail::TrySeqTraits,grpc_composite_call_credentials::GetRequestMetadata::<lambda_1>,grpc_core::MetadataHandle<grpc_core::ServerMetadata>,std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>::PollNonEmpty(void)' being compiled [D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
          with
          [
              T=absl::lts_20211102::StatusOr<grpc_core::ClientMetadataHandle>,
              _Ty=grpc_core::RefCountedPtr<grpc_call_credentials>
          ] (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\security\credentials\composite\composite_credentials.cc)
D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\security\credentials\composite\composite_credentials.cc(61): message : see reference to class template instantiation 'grpc_core::promise_detail::BasicSeqIter<grpc_core::promise_detail::TrySeqTraits,grpc_composite_call_credentials::GetRequestMetadata::<lambda_1>,grpc_core::MetadataHandle<grpc_core::ServerMetadata>,std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>' being compiled [D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
          with
          [
              _Ty=grpc_core::RefCountedPtr<grpc_call_credentials>
          ]
  INTERNAL COMPILER ERROR in 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.32.31326\bin\HostX64\x64\CL.exe'
      Please choose the Technical Support command on the Visual C++
      Help menu, or open the Technical Support help file for more information
cl : command line error D8040: error creating or communicating with child process [D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
  grpc_csharp_plugin.vcxproj -> D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\Release\grpc_csharp_plugin.exe
  grpc_node_plugin.vcxproj -> D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\Release\grpc_node_plugin.exe
  grpc_objective_c_plugin.vcxproj -> D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\Release\grpc_objective_c_plugin.exe
  grpc_php_plugin.vcxproj -> D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\Release\grpc_php_plugin.exe
  grpc_python_plugin.vcxproj -> D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\Release\grpc_python_plugin.exe
  grpc_ruby_plugin.vcxproj -> D:\cmake-toolset\ceb3605\grpc-v1.48.0\windows-amd64-msvc\Release\grpc_ruby_plugin.exe
```

```
D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc(118,13): error C2679: binary '==': no operator found which takes a right-hand operand of type 'int' (or there is no acceptable conversion) [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src/core/lib/slice/slice_internal.h(101,13): message : could be 'bool operator ==(const grpc_slice &,const grpc_slice &)' (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc) [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\shared\guiddef.h(192,15): message : or       'bool operator ==(const GUID &,const GUID &)' (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc) [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.32.31326\include\variant(1682,27): message : or       'bool std::operator ==(std::monostate,std::monostate) noexcept' [found using argument-dependent lookup] (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc) [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\shared\guiddef.h(192,15): message : or 'bool operator ==(const GUID &,const GUID &)' [synthesized expression 'y == x'] (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc) [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.32.31326\include\variant(1682,27): message : or 'bool std::operator ==(std::monostate,std::monostate) noexcept' [synthesized expression 'y == x'] (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc) [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src/core/lib/slice/slice_internal.h(101,13): message : or 'bool operator ==(const grpc_slice &,const grpc_slice &)' [synthesized expression 'y == x'] (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc) [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.32.31326\include\xstring(1740,27): message : or       'bool std::operator ==<char,std::char_traits<char>,2>(const std::basic_string_view<char,std::char_traits<char>>,const std::basic_string_view<char,std::char_traits<char>>) noexcept' (compiling source file D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc) [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
D:\a\cmake-toolset\cmake-toolset\test\third_party\packages\grpc-v1.48.0\src\core\lib\iomgr\resolve_address_windows.cc(118,13): message : while trying to match the argument list '(absl::lts_20211102::string_view, int)' [D:\cmake-toolset\fd370a7\grpc-v1.48.0\windows-amd64-msvc\grpc_unsecure.vcxproj]
```

### Anything else we should know about your project / environment?

https://github.com/atframework/cmake-toolset/runs/7604851855
https://github.com/atframework/cmake-toolset/runs/7624863648

--- a/BUILD
+++ b/BUILD
@@ -1478,6 +1478,7 @@ grpc_cc_library(
         "basic_seq",
         "gpr_platform",
         "poll",
+        "promise_like",
     ],
 )
 
@@ -1496,6 +1497,7 @@ grpc_cc_library(
         "basic_seq",
         "gpr_platform",
         "poll",
+        "promise_like",
         "promise_status",
     ],
 )
@@ -6018,7 +6020,6 @@ grpc_cc_library(
         "activity",
         "arena",
         "arena_promise",
-        "basic_seq",
         "channel_args",
         "channel_fwd",
         "closure",

--- a/src/core/lib/iomgr/resolve_address_windows.cc
+++ b/src/core/lib/iomgr/resolve_address_windows.cc
@@ -119,7 +119,7 @@ NativeDNSResolver::LookupHostnameBlocking(absl::string_view name,
     goto done;
   }
   if (port.empty()) {
-    if (default_port == NULL) {
+    if (default_port.empty()) {
       error = GRPC_ERROR_CREATE_FROM_CPP_STRING(
           absl::StrFormat("no port in name '%s'", name));
       goto done;

--- a/src/core/lib/promise/detail/basic_seq.h
+++ b/src/core/lib/promise/detail/basic_seq.h
@@ -422,10 +422,11 @@ class BasicSeq {
 // As above, but models a sequence of unknown size
 // At each element, the accumulator A and the current value V is passed to some
 // function of type IterTraits::Factory as f(V, IterTraits::Argument); f is
-// expected to return a promise that resolves to Traits::WrappedType.
-template <class Traits, typename IterTraits>
+// expected to return a promise that resolves to IterTraits::Wrapped.
+template <class IterTraits>
 class BasicSeqIter {
  private:
+  using Traits = typename IterTraits::Traits;
   using Iter = typename IterTraits::Iter;
   using Factory = typename IterTraits::Factory;
   using Argument = typename IterTraits::Argument;

--- a/src/core/lib/promise/detail/basic_seq.h
+++ b/src/core/lib/promise/detail/basic_seq.h
@@ -259,16 +259,6 @@ class BasicSeq {
     return &final_promise_;
   }
 
-  // Get the Wrapped type of factory F
-  template <typename F, typename Arg, typename Iter>
-  struct BasicSeqIterTraits {
-    using IterValue = decltype(*std::declval<Iter>());
-    using StateCreated = decltype(std::declval<F>()(std::declval<IterValue>(),
-                                                    std::declval<Arg>()));
-    using State = PromiseLike<StateCreated>;
-    using Wrapped = typename State::Result;
-  };
-
   // Callable to advance the state to the next one after I given the result from
   // state I.
   template <char I>
@@ -431,19 +421,21 @@ class BasicSeq {
 
 // As above, but models a sequence of unknown size
 // At each element, the accumulator A and the current value V is passed to some
-// function of type F as f(V, A); f is expected to return a promise that
-// resolves to Traits::WrappedType.
-template <class Traits, typename F, typename Arg, typename Iter>
+// function of type IterTraits::Factory as f(V, IterTraits::Argument); f is
+// expected to return a promise that resolves to Traits::WrappedType.
+template <class Traits, typename IterTraits>
 class BasicSeqIter {
  private:
-  using IterValue = decltype(*std::declval<Iter>());
-  using StateCreated = decltype(std::declval<F>()(std::declval<IterValue>(),
-                                                  std::declval<Arg>()));
-  using State = PromiseLike<StateCreated>;
-  using Wrapped = typename State::Result;
+  using Iter = typename IterTraits::Iter;
+  using Factory = typename IterTraits::Factory;
+  using Argument = typename IterTraits::Argument;
+  using IterValue = typename IterTraits::IterValue;
+  using StateCreated = typename IterTraits::StateCreated;
+  using State = typename IterTraits::State;
+  using Wrapped = typename IterTraits::Wrapped;
 
  public:
-  BasicSeqIter(Iter begin, Iter end, F f, Arg arg)
+  BasicSeqIter(Iter begin, Iter end, Factory f, Argument arg)
       : cur_(begin), end_(end), f_(std::move(f)) {
     if (cur_ == end_) {
       Construct(&result_, std::move(arg));
@@ -510,10 +502,10 @@ class BasicSeqIter {
 
   Iter cur_;
   const Iter end_;
-  GPR_NO_UNIQUE_ADDRESS F f_;
+  GPR_NO_UNIQUE_ADDRESS Factory f_;
   union {
     GPR_NO_UNIQUE_ADDRESS State state_;
-    GPR_NO_UNIQUE_ADDRESS Arg result_;
+    GPR_NO_UNIQUE_ADDRESS Argument result_;
   };
 };
 

--- a/src/core/lib/promise/detail/basic_seq.h
+++ b/src/core/lib/promise/detail/basic_seq.h
@@ -259,6 +259,16 @@ class BasicSeq {
     return &final_promise_;
   }
 
+  // Get the Wrapped type of factory F
+  template <typename F, typename Arg, typename Iter>
+  struct BasicSeqIterTraits {
+    using IterValue = decltype(*std::declval<Iter>());
+    using StateCreated = decltype(std::declval<F>()(std::declval<IterValue>(),
+                                                    std::declval<Arg>()));
+    using State = PromiseLike<StateCreated>;
+    using Wrapped = typename State::Result;
+  };
+
   // Callable to advance the state to the next one after I given the result from
   // state I.
   template <char I>
@@ -423,8 +433,7 @@ class BasicSeq {
 // At each element, the accumulator A and the current value V is passed to some
 // function of type F as f(V, A); f is expected to return a promise that
 // resolves to Traits::WrappedType.
-template <template <typename Wrapped> class Traits, typename F, typename Arg,
-          typename Iter>
+template <class Traits, typename F, typename Arg, typename Iter>
 class BasicSeqIter {
  private:
   using IterValue = decltype(*std::declval<Iter>());
@@ -484,7 +493,7 @@ class BasicSeqIter {
   Poll<Wrapped> PollNonEmpty() {
     Poll<Wrapped> r = state_();
     if (absl::holds_alternative<Pending>(r)) return r;
-    return Traits<Wrapped>::template CheckResultAndRunNext<Wrapped>(
+    return Traits::template CheckResultAndRunNext<Wrapped>(
         std::move(absl::get<Wrapped>(r)), [this](Wrapped arg) -> Poll<Wrapped> {
           auto next = cur_;
           ++next;
@@ -494,7 +503,7 @@ class BasicSeqIter {
           cur_ = next;
           state_.~State();
           Construct(&state_,
-                    Traits<Wrapped>::CallSeqFactory(f_, *cur_, std::move(arg)));
+                    Traits::template CallSeqFactory(f_, *cur_, std::move(arg)));
           return PollNonEmpty();
         });
   }

--- a/src/core/lib/promise/seq.h
+++ b/src/core/lib/promise/seq.h
@@ -75,11 +75,16 @@ F Seq(F functor) {
 //   }
 //   return argument;
 template <typename Iter, typename Factory, typename Argument>
-promise_detail::BasicSeqIter<promise_detail::SeqTraits, Factory, Argument, Iter>
+promise_detail::BasicSeqIter<
+    promise_detail::SeqTraits<typename promise_detail::BasicSeqIterTraits<
+        Factory, Argument, Iter>::Wrapped>,
+    Factory, Argument, Iter>
 SeqIter(Iter begin, Iter end, Argument argument, Factory factory) {
-  return promise_detail::BasicSeqIter<promise_detail::SeqTraits, Factory,
-                                      Argument, Iter>(
-      begin, end, std::move(factory), std::move(argument));
+  return promise_detail::BasicSeqIter<
+      promise_detail::SeqTraits<typename promise_detail::BasicSeqIterTraits<
+          Factory, Argument, Iter>::Wrapped>,
+      Factory, Argument, Iter>(begin, end, std::move(factory),
+                               std::move(argument));
 }
 
 }  // namespace grpc_core

--- a/src/core/lib/promise/seq.h
+++ b/src/core/lib/promise/seq.h
@@ -20,6 +20,7 @@
 #include <utility>
 
 #include "src/core/lib/promise/detail/basic_seq.h"
+#include "src/core/lib/promise/detail/promise_like.h"
 #include "src/core/lib/promise/poll.h"
 
 namespace grpc_core {

--- a/src/core/lib/promise/seq.h
+++ b/src/core/lib/promise/seq.h
@@ -50,6 +50,25 @@ struct SeqTraits {
 template <typename... Fs>
 using Seq = BasicSeq<SeqTraits, Fs...>;
 
+template <typename I, typename F, typename Arg>
+struct SeqIterTraits {
+  using Iter = I;
+  using Factory = F;
+  using Argument = Arg;
+  using IterValue = decltype(*std::declval<Iter>());
+  using StateCreated = decltype(std::declval<F>()(std::declval<IterValue>(),
+                                                  std::declval<Arg>()));
+  using State = PromiseLike<StateCreated>;
+  using Wrapped = typename State::Result;
+};
+
+template <typename Iter, typename Factory, typename Argument>
+struct SeqIterResultTraits {
+  using IterTraits = SeqIterTraits<Iter, Factory, Argument>;
+  using Traits = SeqTraits<typename IterTraits::Wrapped>;
+  using Result = BasicSeqIter<Traits, IterTraits>;
+};
+
 }  // namespace promise_detail
 
 // Sequencing combinator.
@@ -75,16 +94,11 @@ F Seq(F functor) {
 //   }
 //   return argument;
 template <typename Iter, typename Factory, typename Argument>
-promise_detail::BasicSeqIter<
-    promise_detail::SeqTraits<typename promise_detail::BasicSeqIterTraits<
-        Factory, Argument, Iter>::Wrapped>,
-    Factory, Argument, Iter>
+typename promise_detail::SeqIterResultTraits<Iter, Factory, Argument>::Result
 SeqIter(Iter begin, Iter end, Argument argument, Factory factory) {
-  return promise_detail::BasicSeqIter<
-      promise_detail::SeqTraits<typename promise_detail::BasicSeqIterTraits<
-          Factory, Argument, Iter>::Wrapped>,
-      Factory, Argument, Iter>(begin, end, std::move(factory),
-                               std::move(argument));
+  using Result = typename promise_detail::SeqIterResultTraits<Iter, Factory,
+                                                              Argument>::Result;
+  return Result(begin, end, std::move(factory), std::move(argument));
 }
 
 }  // namespace grpc_core

--- a/src/core/lib/promise/seq.h
+++ b/src/core/lib/promise/seq.h
@@ -60,13 +60,14 @@ struct SeqIterTraits {
                                                   std::declval<Arg>()));
   using State = PromiseLike<StateCreated>;
   using Wrapped = typename State::Result;
+
+  using Traits = SeqTraits<Wrapped>;
 };
 
 template <typename Iter, typename Factory, typename Argument>
 struct SeqIterResultTraits {
   using IterTraits = SeqIterTraits<Iter, Factory, Argument>;
-  using Traits = SeqTraits<typename IterTraits::Wrapped>;
-  using Result = BasicSeqIter<Traits, IterTraits>;
+  using Result = BasicSeqIter<IterTraits>;
 };
 
 }  // namespace promise_detail

--- a/src/core/lib/promise/try_seq.h
+++ b/src/core/lib/promise/try_seq.h
@@ -115,6 +115,25 @@ using TrySeqTraits = TrySeqTraitsWithSfinae<T>;
 template <typename... Fs>
 using TrySeq = BasicSeq<TrySeqTraits, Fs...>;
 
+template <typename I, typename F, typename Arg>
+struct TrySeqIterTraits {
+  using Iter = I;
+  using Factory = F;
+  using Argument = Arg;
+  using IterValue = decltype(*std::declval<Iter>());
+  using StateCreated = decltype(std::declval<F>()(std::declval<IterValue>(),
+                                                  std::declval<Arg>()));
+  using State = PromiseLike<StateCreated>;
+  using Wrapped = typename State::Result;
+};
+
+template <typename Iter, typename Factory, typename Argument>
+struct TrySeqIterResultTraits {
+  using IterTraits = TrySeqIterTraits<Iter, Factory, Argument>;
+  using Traits = TrySeqTraits<typename IterTraits::Wrapped>;
+  using Result = BasicSeqIter<Traits, IterTraits>;
+};
+
 }  // namespace promise_detail
 
 // Try a sequence of operations.
@@ -143,16 +162,12 @@ promise_detail::TrySeq<Functors...> TrySeq(Functors... functors) {
 //   }
 //   return argument;
 template <typename Iter, typename Factory, typename Argument>
-promise_detail::BasicSeqIter<
-    promise_detail::TrySeqTraits<typename promise_detail::BasicSeqIterTraits<
-        Factory, Argument, Iter>::Wrapped>,
-    Factory, Argument, Iter>
+typename promise_detail::TrySeqIterResultTraits<Iter, Factory, Argument>::Result
 TrySeqIter(Iter begin, Iter end, Argument argument, Factory factory) {
-  return promise_detail::BasicSeqIter<
-      promise_detail::TrySeqTraits<typename promise_detail::BasicSeqIterTraits<
-          Factory, Argument, Iter>::Wrapped>,
-      Factory, Argument, Iter>(begin, end, std::move(factory),
-                               std::move(argument));
+  using Result =
+      typename promise_detail::TrySeqIterResultTraits<Iter, Factory,
+                                                      Argument>::Result;
+  return Result(begin, end, std::move(factory), std::move(argument));
 }
 
 }  // namespace grpc_core

--- a/src/core/lib/promise/try_seq.h
+++ b/src/core/lib/promise/try_seq.h
@@ -24,6 +24,7 @@
 #include "absl/status/statusor.h"
 
 #include "src/core/lib/promise/detail/basic_seq.h"
+#include "src/core/lib/promise/detail/promise_like.h"
 #include "src/core/lib/promise/detail/status.h"
 #include "src/core/lib/promise/poll.h"
 

--- a/src/core/lib/promise/try_seq.h
+++ b/src/core/lib/promise/try_seq.h
@@ -125,13 +125,14 @@ struct TrySeqIterTraits {
                                                   std::declval<Arg>()));
   using State = PromiseLike<StateCreated>;
   using Wrapped = typename State::Result;
+
+  using Traits = TrySeqTraits<Wrapped>;
 };
 
 template <typename Iter, typename Factory, typename Argument>
 struct TrySeqIterResultTraits {
   using IterTraits = TrySeqIterTraits<Iter, Factory, Argument>;
-  using Traits = TrySeqTraits<typename IterTraits::Wrapped>;
-  using Result = BasicSeqIter<Traits, IterTraits>;
+  using Result = BasicSeqIter<IterTraits>;
 };
 
 }  // namespace promise_detail

--- a/src/core/lib/promise/try_seq.h
+++ b/src/core/lib/promise/try_seq.h
@@ -143,12 +143,16 @@ promise_detail::TrySeq<Functors...> TrySeq(Functors... functors) {
 //   }
 //   return argument;
 template <typename Iter, typename Factory, typename Argument>
-promise_detail::BasicSeqIter<promise_detail::TrySeqTraits, Factory, Argument,
-                             Iter>
+promise_detail::BasicSeqIter<
+    promise_detail::TrySeqTraits<typename promise_detail::BasicSeqIterTraits<
+        Factory, Argument, Iter>::Wrapped>,
+    Factory, Argument, Iter>
 TrySeqIter(Iter begin, Iter end, Argument argument, Factory factory) {
-  return promise_detail::BasicSeqIter<promise_detail::TrySeqTraits, Factory,
-                                      Argument, Iter>(
-      begin, end, std::move(factory), std::move(argument));
+  return promise_detail::BasicSeqIter<
+      promise_detail::TrySeqTraits<typename promise_detail::BasicSeqIterTraits<
+          Factory, Argument, Iter>::Wrapped>,
+      Factory, Argument, Iter>(begin, end, std::move(factory),
+                               std::move(argument));
 }
 
 }  // namespace grpc_core

--- a/src/core/lib/security/credentials/composite/composite_credentials.cc
+++ b/src/core/lib/security/credentials/composite/composite_credentials.cc
@@ -31,7 +31,6 @@
 
 #include "src/core/lib/debug/trace.h"
 #include "src/core/lib/gprpp/ref_counted_ptr.h"
-#include "src/core/lib/promise/detail/basic_seq.h"
 #include "src/core/lib/promise/poll.h"
 #include "src/core/lib/promise/try_seq.h"
 #include "src/core/lib/surface/api_trace.h"

